### ISSUE TITLE
How to pass arguments to rundeck job though url?

### ISSUE THREAD

**Issue by sduggisetty** [rundeck/rundeck#1176]
Hi,

Am able to call rundeck job through url, but facing problem to pass dynamic parameters through url. 

I am trying to pass parameter though url as below.

"http://<server-name>:4440/api/1/job/<job-id>/run?authtoken=<authentication-token>&argString=-arg1=val1"

could someone please help me.

-Subbu

**Comment by mumblez** [comment]
For the arguments you'd use variable format ${option.variable} in your url.
Not sure what you're trying to do though, if you already have a job in rundeck and trying to call that from another job you could use a job reference!

Then you can pass arguments to the job using ${option.value} syntax, e.g. if your referenced job had an argument called "hostremote" you can do:

-hostremote ${option.hostlocal} -arg2 ${option.arg2} -arg3 ${option.arg3} (in the arguments field)

-hostremote = the referenced job argument
-hostlocal = argument (option) created that's calling the referenced job

Hope this makes sense.

These sort's of question might be better served if you use the google groups - https://groups.google.com/forum/#!forum/rundeck-discuss

**Comment by sduggisetty** [comment]
@mumblez 
I am trying to call rundeck job from jenkins job through http request.

Currently when i run jenkins job (through url) it is able to trigger my rundeck job, but unable to pass parameters to rundeck job.

I need to pass build number to my rundeck job, so that my rundeck job can start deployment based on the build number.

-Subbu

**Comment by mumblez** [comment]
I don't know how to do via the rest api but I have the rundeck plugin installed for Jenkins, it's a post build step, you're able to call a job by it's uuid and also pass in options (key=value per line), maybe this will help?

**Comment by sduggisetty** [comment]
@mumblez 
yes, i am also able to pass parameters using rundeck plugin.

The problem is, i have two rundeck servers. i need to use both the rundecks and do the deployments.
Using rundeck plugin, i am able to configure single rundeck and work with it successfully. But i need to call second rundeck job too through passing few parameters.

-Subbu

**Comment by mumblez** [comment]
I found this by @ahonor - https://gist.github.com/ahonor/1006584
Maybe this will help, see near bottom for arguments format, e.g.
-arg1 value -arg2 value2

The key bit is that the data is url encoded so the spaces get correctly encoded. Hope this helps

**Comment by sduggisetty** [comment]
Got some idea. Thank you. :-)
@mumblez 
It seems we are using "curl" command in the given link, i am trying to pass parameter using via HTTP protocol (without using curl).

i am able to pass parameter using this command.
curl --data-urlencode "argString=-arg1 val1" http://<rundeck-url>/api/1/job/job-id/run?authtoken=<authorization token>

What i am trying is... need to pass parameters without using curl like..(not working)
http://<rundeck-url>/api/1/job/job-id/run?authtoken=<authorization token>&arg1=val1

-Subbu

**Comment by mumblez** [comment]
np, haven't tried it but you might be able to do it if you encode the values (e.g. if it has spaces replace with +), encoder tool - http://www.url-encode-decode.com/

**Comment by gschueler** [comment]
yes you need to urlencode the argString value: `argString=-arg1+val1`

**Comment by sduggisetty** [comment]
Thank you for all your help

**Comment by danifr** [comment]
Hi @sduggisetty,

 can you share how did you manage to do it? 
I basically want to offer my users and URL like "Click on this link" (and the link runs a Rundeck job behind the scenes...)

Thanks!

**Comment by sduggisetty** [comment]
Hi @danifr 

We can pass multiple arguments through rundeck url as like below. I have managed using below url.

curl --data-urlencode "argString=-arg1 value1 -arg2 value2" http://rundeckhost:4440/api/1/job/$rundeck_job_id/run?authtoken=<auth token>

**Comment by danifr** [comment]
Hi @sduggisetty,

I see. I was expecting something I could insert as a HTML link in an email.
Thanks anyway for your help :)

**Comment by Mykolaichenko** [comment]
Hello, here is my implementation 

```
class RundeckJob():
    def get_full_url(self):
        return 'http://rundeck.api.com/api/12/job/' + self.job_id + '/run?authtoken=' + RUNDECK_TOKEN

    def make_api_call(self):
        opener = urllib2.build_opener(urllib2.HTTPHandler)
        request = urllib2.Request(self.get_full_url())
        try:
            opener.open(request, timeout=30)
        except urllib2.HTTPError as err:
            if err.code == 409:
                return 'Already running'
        except Exception as e:
            logging.error('General exception: %s', e)
            return False

    def make_post_api_call(self, post):
        """

        :param post: {'argString':'-version blabla'}
        :return:
        """
        post_query = urllib.urlencode(post)
        req = urllib2.Request('http://rundeck.api.com/api/12/job/' + self.job_id + '/run', data=post_query,
                              headers={'X-RunDeck-Auth-Token': RUNDECK_TOKEN})
        try:
            urllib2.urlopen(req).read()
        except urllib2.HTTPError as err:
            if err.code == 409:
                return 'Already running'
        except Exception as e:
            logging.error('General exception: %s', e)
            return False

```

**Comment by danifr** [comment]
Hi @Mykolaichenko,

they officially implemented this feature in v2.7.0 Works perfectly!
https://github.com/rundeck/rundeck/pull/2209
